File: JustInTimeAlerts/Services/IcsParserService.cs
using System.Net;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using Ical.Net;
using Ical.Net.CalendarComponents;
using Ical.Net.DataTypes;
using JustInTimeAlerts.Models;

namespace JustInTimeAlerts.Services;

/// <summary>
/// Fetches and parses ICS calendars from a URL or a local file path,
/// returning strongly-typed <see cref="MeetingEvent"/> objects.
/// </summary>
/// <remarks>
/// Network-efficiency strategy (prevents tight-loop DOS behaviour):
/// <list type="bullet">
///   <item>URLs are re-fetched at most once every <see cref="MinFetchInterval"/> (15 min).</item>
///   <item>Every URL request sends <c>If-None-Match</c> / <c>If-Modified-Since</c> headers.
///         A <c>304 Not Modified</c> response returns the in-memory cache instantly.</item>
///   <item>Even on a 200 response the ICS is only re-parsed when the SHA-256
///         content hash differs from the previously cached value.</item>
///   <item>Consecutive fetch failures trigger exponential back-off (1 min → 5 min → 15 min)
///         so a broken feed never produces an infinite retry loop.</item>
/// </list>
/// </remarks>
public class IcsParserService
{
    // -----------------------------------------------------------------------
    // Constants
    // -----------------------------------------------------------------------

    /// <summary>Minimum time between actual HTTP fetches for the same URL.</summary>
    public static readonly TimeSpan MinFetchInterval = TimeSpan.FromMinutes(15);

    private static readonly TimeSpan[] BackoffSteps =
    [
        TimeSpan.FromMinutes(1),
        TimeSpan.FromMinutes(5),
        TimeSpan.FromMinutes(15),
    ];

    // -----------------------------------------------------------------------
    // Per-source in-memory cache
    // -----------------------------------------------------------------------

    private sealed class SourceCache
    {
        public string?                    ETag              { get; set; }
        public string?                    LastModified       { get; set; }
        public string?                    ContentHash        { get; set; }
        public IReadOnlyList<MeetingEvent> Events            { get; set; } = Array.Empty<MeetingEvent>();
        public DateTime                   LastFetchTime      { get; set; } = DateTime.MinValue;
        public int                        ConsecutiveFailures { get; set; }
        public DateTime                   BackoffUntil       { get; set; } = DateTime.MinValue;

        // For local files: track the last write time so we only re-parse on change.
        public DateTime                   FileLastWrite      { get; set; } = DateTime.MinValue;
    }

    private readonly Dictionary<string, SourceCache> _cache = new(StringComparer.OrdinalIgnoreCase);
    private readonly object _cacheLock = new();

    // -----------------------------------------------------------------------

    private readonly HttpClient  _httpClient;
    private readonly DebugLogService _log;

    public IcsParserService(HttpClient httpClient, DebugLogService log)
    {
        _httpClient = httpClient;
        _log = log;
    }

    // -----------------------------------------------------------------------
    // Public API
    // -----------------------------------------------------------------------

    /// <summary>
    /// Wipes all in-memory cache entries (or a single entry keyed by
    /// <paramref name="cacheKey"/>) so that the next
    /// <see cref="GetEventsAsync"/> call performs an unconditional HTTP fetch
    /// and full re-parse, bypassing the minimum re-fetch interval, back-off,
    /// conditional-request headers, and content-hash dedup.
    /// </summary>
    /// <param name="cacheKey">
    /// The URL or file path that identifies the source to invalidate.
    /// Pass <c>null</c> (the default) to invalidate every source at once.
    /// </param>
    public void InvalidateCache(string? cacheKey = null)
    {
        lock (_cacheLock)
        {
            if (cacheKey is null)
            {
                foreach (var entry in _cache.Values)
                    ResetEntry(entry);

                _log.Log($"ICS cache: all {_cache.Count} source(s) invalidated for force-sync.");
            }
            else if (_cache.TryGetValue(cacheKey, out var entry))
            {
                ResetEntry(entry);
                _log.Log($"ICS cache: '{cacheKey}' invalidated for force-sync.");
            }
        }

        static void ResetEntry(SourceCache e)
        {
            e.LastFetchTime       = DateTime.MinValue;
            e.BackoffUntil        = DateTime.MinValue;
            e.ContentHash         = null;
            e.ETag                = null;
            e.LastModified        = null;
            e.FileLastWrite       = DateTime.MinValue;
            e.ConsecutiveFailures = 0;
        }
    }

    /// <summary>
    /// Loads events from the given <paramref name="source"/>.
    /// Returns cached events when no network fetch is needed.
    /// Returns an empty list on unrecoverable failure so callers stay clean.
    /// </summary>
    public async Task<IReadOnlyList<MeetingEvent>> GetEventsAsync(
        CalendarSource source,
        CancellationToken cancellationToken = default)
    {
        if (!string.IsNullOrWhiteSpace(source.Url))
            return await FetchFromUrlAsync(source, cancellationToken).ConfigureAwait(false);

        if (!string.IsNullOrWhiteSpace(source.FilePath) && File.Exists(source.FilePath))
            return await ReadFromFileAsync(source, cancellationToken).ConfigureAwait(false);

        _log.Log("ICS source has no URL or valid file path — skipping.");
        return Array.Empty<MeetingEvent>();
    }

    // -----------------------------------------------------------------------
    // URL path
    // -----------------------------------------------------------------------

    private async Task<IReadOnlyList<MeetingEvent>> FetchFromUrlAsync(
        CalendarSource source,
        CancellationToken cancellationToken)
    {
        var cacheKey = source.Url!;
        SourceCache entry;
        lock (_cacheLock)
        {
            if (!_cache.TryGetValue(cacheKey, out entry!))
            {
                entry = new SourceCache();
                _cache[cacheKey] = entry;
            }
        }

        var now = DateTime.UtcNow;

        // ── Back-off guard ──────────────────────────────────────────────────
        if (now < entry.BackoffUntil)
        {
            _log.Log($"ICS [{source.DisplayName}]: in back-off until {entry.BackoffUntil:HH:mm:ss} UTC — returning cache.");
            return entry.Events;
        }

        // ── Minimum re-fetch interval ───────────────────────────────────────
        if (entry.Events.Count > 0 && (now - entry.LastFetchTime) < MinFetchInterval)
        {
            _log.Log($"ICS [{source.DisplayName}]: cache fresh (last fetch {(now - entry.LastFetchTime).TotalMinutes:F1} min ago) — skipping HTTP.");
            return entry.Events;
        }

        // ── Build conditional request ───────────────────────────────────────
        using var request = new HttpRequestMessage(HttpMethod.Get, source.Url);

        if (!string.IsNullOrEmpty(entry.ETag))
            request.Headers.IfNoneMatch.Add(new EntityTagHeaderValue(entry.ETag, isWeak: true));
        else if (!string.IsNullOrEmpty(entry.LastModified) &&
                 DateTimeOffset.TryParse(entry.LastModified, out var lmParsed))
            request.Headers.IfModifiedSince = lmParsed;

        HttpResponseMessage response;
        try
        {
            _log.Log($"ICS [{source.DisplayName}]: HTTP GET (ETag: {entry.ETag ?? "none"})");
            response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            return HandleFetchFailure(entry, source.DisplayName, ex);
        }

        using (response)
        {
            // ── 304 Not Modified ────────────────────────────────────────────
            if (response.StatusCode == HttpStatusCode.NotModified)
            {
                entry.ConsecutiveFailures = 0;
                entry.LastFetchTime = now;
                _log.Log($"ICS [{source.DisplayName}]: 304 Not Modified — using {entry.Events.Count} cached event(s).");
                return entry.Events;
            }

            // ── Non-success ─────────────────────────────────────────────────
            if (!response.IsSuccessStatusCode)
            {
                return HandleFetchFailure(entry, source.DisplayName,
                    new HttpRequestException($"HTTP {(int)response.StatusCode} {response.ReasonPhrase}"));
            }

            // ── Read body ───────────────────────────────────────────────────
            string icsContent;
            try
            {
                icsContent = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                return HandleFetchFailure(entry, source.DisplayName, ex);
            }

            // ── Content-hash dedup (skip re-parse if identical) ─────────────
            var hash = ComputeHash(icsContent);
            if (hash == entry.ContentHash)
            {
                entry.ConsecutiveFailures = 0;
                entry.LastFetchTime = now;
                // Refresh cache-control metadata in case ETag changed.
                CaptureResponseHeaders(response, entry);
                _log.Log($"ICS [{source.DisplayName}]: 200 but content unchanged — using {entry.Events.Count} cached event(s).");
                return entry.Events;
            }

            // ── Parse new content ───────────────────────────────────────────
            _log.Log($"ICS [{source.DisplayName}]: new content ({icsContent.Length} chars) — parsing.");
            var parsed = ParseIcsContent(icsContent, source.Id);
            _log.Log($"ICS [{source.DisplayName}]: parsed {parsed.Count} event(s).");

            entry.ContentHash          = hash;
            entry.Events               = parsed;
            entry.LastFetchTime        = now;
            entry.ConsecutiveFailures  = 0;
            entry.BackoffUntil         = DateTime.MinValue;
            CaptureResponseHeaders(response, entry);

            return parsed;
        }
    }

    // -----------------------------------------------------------------------
    // File path
    // -----------------------------------------------------------------------

    private async Task<IReadOnlyList<MeetingEvent>> ReadFromFileAsync(
        CalendarSource source,
        CancellationToken cancellationToken)
    {
        var cacheKey = source.FilePath!;
        SourceCache entry;
        lock (_cacheLock)
        {
            if (!_cache.TryGetValue(cacheKey, out entry!))
            {
                entry = new SourceCache();
                _cache[cacheKey] = entry;
            }
        }

        var lastWrite = File.GetLastWriteTimeUtc(source.FilePath!);
        if (entry.Events.Count > 0 && lastWrite == entry.FileLastWrite)
        {
            _log.Log($"ICS [{source.DisplayName}]: file unchanged — using {entry.Events.Count} cached event(s).");
            return entry.Events;
        }

        string icsContent;
        try
        {
            _log.Log($"ICS [{source.DisplayName}]: reading file.");
            icsContent = await File.ReadAllTextAsync(source.FilePath!, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _log.Log($"ERROR reading ICS file: {ex.GetType().Name}: {ex.Message}");
            return entry.Events; // return stale cache rather than empty
        }

        var parsed = ParseIcsContent(icsContent, source.Id);
        _log.Log($"ICS [{source.DisplayName}]: parsed {parsed.Count} event(s) from file.");

        entry.Events        = parsed;
        entry.FileLastWrite = lastWrite;
        entry.ContentHash   = ComputeHash(icsContent);

        return parsed;
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    private IReadOnlyList<MeetingEvent> HandleFetchFailure(
        SourceCache entry, string displayName, Exception ex)
    {
        entry.ConsecutiveFailures++;
        var stepIdx    = Math.Min(entry.ConsecutiveFailures - 1, BackoffSteps.Length - 1);
        var backoff    = BackoffSteps[stepIdx];
        entry.BackoffUntil = DateTime.UtcNow + backoff;

        _log.Log($"ERROR fetching ICS [{displayName}] (failure #{entry.ConsecutiveFailures}): " +
                 $"{ex.GetType().Name}: {ex.Message}. " +
                 $"Back-off for {backoff.TotalMinutes:F0} min.");

        return entry.Events; // return stale cache rather than empty
    }

    private static void CaptureResponseHeaders(HttpResponseMessage response, SourceCache entry)
    {
        if (response.Headers.ETag != null)
            entry.ETag = response.Headers.ETag.Tag;

        if (response.Content.Headers.LastModified.HasValue)
            entry.LastModified = response.Content.Headers.LastModified.Value.ToString("R");
    }

    private static string ComputeHash(string content)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(content));
        return Convert.ToHexString(bytes);
    }

    /// <summary>
    /// Removes invalid combinations from RRULE lines where both COUNT and UNTIL are present.
    /// RFC 5545 §3.3.10 forbids supplying both; real-world feeds (Outlook, Google) do it anyway.
    /// We keep UNTIL and drop COUNT, since UNTIL is more precise.
    /// </summary>
    private static string SanitizeIcsContent(string icsContent)
    {
        // Step 1: unfold RFC 5545 line continuations (CRLF/LF followed by SPACE or TAB)
        // so that each logical property is on a single string we can inspect.
        var unfolded = System.Text.RegularExpressions.Regex.Replace(
            icsContent, @"\r?\n[ \t]", string.Empty);

        // Step 2: for every logical RRULE line that contains both UNTIL= and COUNT=,
        // strip COUNT= (keeping UNTIL= as it is more precise).
        var sanitized = System.Text.RegularExpressions.Regex.Replace(
            unfolded,
            @"(?im)^(RRULE[^:]*:[^\r\n]*UNTIL=[^\r\n]*)(?:;COUNT=\d+|COUNT=\d+;?)([^\r\n]*)$",
            "$1$2");

        // Also handle COUNT before UNTIL
        sanitized = System.Text.RegularExpressions.Regex.Replace(
            sanitized,
            @"(?im)^(RRULE[^:]*:(?:[^\r\n]*?))COUNT=\d+;([^\r\n]*UNTIL=[^\r\n]*)$",
            "$1$2");

        return sanitized;
    }

    /// <summary>Parses raw ICS text and returns a list of <see cref="MeetingEvent"/>.</summary>
    public IReadOnlyList<MeetingEvent> ParseIcsContent(string icsContent, Guid calendarSourceId)
    {
        if (string.IsNullOrWhiteSpace(icsContent))
            return Array.Empty<MeetingEvent>();

        try
        {
            icsContent = SanitizeIcsContent(icsContent);
            var calendar = Calendar.Load(icsContent);
            var events = new List<MeetingEvent>();

            foreach (CalendarEvent e in calendar.Events)
            {
                if (e.Start == null)
                    continue;

                var uid = string.IsNullOrWhiteSpace(e.Uid)
                    ? Guid.NewGuid().ToString()
                    : e.Uid;

                // For recurring events expand occurrences over the next 30 days.
                if (e.RecurrenceRules?.Count > 0)
                {
                    var rangeStart = new CalDateTime(DateTime.UtcNow);
                    var rangeEnd   = new CalDateTime(DateTime.UtcNow.AddDays(30));
                    var occurrences = e.GetOccurrences(rangeStart)
                        .TakeWhile(occ => occ.Period.StartTime <= rangeEnd);
                    foreach (var occ in occurrences)
                    {
                        events.Add(new MeetingEvent
                        {
                            Uid = $"{uid}_{occ.Period.StartTime.AsUtc:yyyyMMddHHmmss}",
                            Title = e.Summary ?? "(No Title)",
                            Start = occ.Period.StartTime.AsUtc,
                            End = occ.Period.EndTime?.AsUtc ?? occ.Period.StartTime.AsUtc.AddHours(1),
                            Description = e.Description,
                            Location = e.Location,
                            CalendarSourceId = calendarSourceId,
                        });
                    }
                }
                else
                {
                    events.Add(new MeetingEvent
                    {
                        Uid = uid,
                        Title = e.Summary ?? "(No Title)",
                        Start = e.Start.AsUtc,
                        End = e.End?.AsUtc ?? e.Start.AsUtc.AddHours(1),
                        Description = e.Description,
                        Location = e.Location,
                        CalendarSourceId = calendarSourceId,
                    });
                }
            }

            return events;
        }
        catch (Exception ex)
        {
            _log.Log($"ERROR parsing ICS content: {ex.GetType().Name}: {ex.Message}");
            return Array.Empty<MeetingEvent>();
        }
    }
}

File: app/src/main/java/com/justintimealerts/services/IcsParserService.kt
package com.justintimealerts.services

import com.justintimealerts.models.CalendarSource
import com.justintimealerts.models.MeetingEvent
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.fortuna.ical4j.data.CalendarBuilder
import net.fortuna.ical4j.model.Calendar
import net.fortuna.ical4j.model.Component
import net.fortuna.ical4j.model.Property
import net.fortuna.ical4j.model.component.VEvent
import net.fortuna.ical4j.model.property.DtEnd
import net.fortuna.ical4j.model.property.DtStart
import net.fortuna.ical4j.model.property.RRule
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.io.StringReader
import java.security.MessageDigest
import java.time.Duration
import java.time.Instant
import java.time.temporal.Temporal
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Fetches and parses ICS calendars from a URL or a local file path,
 * returning strongly-typed [MeetingEvent] objects.
 *
 * Network-efficiency strategy (prevents tight-loop DOS behaviour):
 * - URLs are re-fetched at most once every [MIN_FETCH_INTERVAL_MS] (15 min).
 * - Every URL request sends If-None-Match / If-Modified-Since headers.
 *   A 304 Not Modified response returns the in-memory cache instantly.
 * - Even on a 200 response the ICS is only re-parsed when the SHA-256
 *   content hash differs from the previously cached value.
 * - Consecutive fetch failures trigger exponential back-off (1 min → 5 min → 15 min)
 *   so a broken feed never produces an infinite retry loop.
 */
class IcsParserService(
    private val httpClient: OkHttpClient,
    private val log: DebugLogService
) {
    companion object {
        /** Minimum time between actual HTTP fetches for the same URL. */
        const val MIN_FETCH_INTERVAL_MS: Long = 15 * 60 * 1000L // 15 minutes
        
        private val BACKOFF_STEPS = longArrayOf(
            1 * 60 * 1000L,  // 1 minute
            5 * 60 * 1000L,  // 5 minutes
            15 * 60 * 1000L  // 15 minutes
        )
    }
    
    private data class SourceCache(
        var eTag: String? = null,
        var lastModified: String? = null,
        var contentHash: String? = null,
        var events: List<MeetingEvent> = emptyList(),
        var lastFetchTime: Long = 0L,
        var consecutiveFailures: Int = 0,
        var backoffUntil: Long = 0L,
        var fileLastWrite: Long = 0L
    )
    
    private val cache = ConcurrentHashMap<String, SourceCache>()
    
    /**
     * Wipes all in-memory cache entries (or a single entry keyed by [cacheKey])
     * so that the next [getEvents] call performs an unconditional HTTP fetch
     * and full re-parse.
     */
    fun invalidateCache(cacheKey: String? = null) {
        if (cacheKey == null) {
            cache.values.forEach { resetEntry(it) }
            log.log("ICS cache: all ${cache.size} source(s) invalidated for force-sync.")
        } else {
            cache[cacheKey]?.let {
                resetEntry(it)
                log.log("ICS cache: '$cacheKey' invalidated for force-sync.")
            }
        }
    }
    
    private fun resetEntry(entry: SourceCache) {
        entry.lastFetchTime = 0L
        entry.backoffUntil = 0L
        entry.contentHash = null
        entry.eTag = null
        entry.lastModified = null
        entry.fileLastWrite = 0L
        entry.consecutiveFailures = 0
    }
    
    /**
     * Loads events from the given [source].
     * Returns cached events when no network fetch is needed.
     * Returns an empty list on unrecoverable failure so callers stay clean.
     */
    suspend fun getEvents(source: CalendarSource): List<MeetingEvent> = withContext(Dispatchers.IO) {
        when {
            !source.url.isNullOrBlank() -> fetchFromUrl(source)
            !source.filePath.isNullOrBlank() && File(source.filePath).exists() -> readFromFile(source)
            else -> {
                log.log("ICS source has no URL or valid file path — skipping.")
                emptyList()
            }
        }
    }
    
    private suspend fun fetchFromUrl(source: CalendarSource): List<MeetingEvent> = withContext(Dispatchers.IO) {
        val cacheKey = source.url!!
        val entry = cache.getOrPut(cacheKey) { SourceCache() }
        val now = System.currentTimeMillis()
        
        // Back-off guard
        if (now < entry.backoffUntil) {
            log.log("ICS [${source.displayName}]: in back-off until ${entry.backoffUntil} — returning cache.")
            return@withContext entry.events
        }
        
        // Minimum re-fetch interval
        if (entry.events.isNotEmpty() && (now - entry.lastFetchTime) < MIN_FETCH_INTERVAL_MS) {
            val minAgo = (now - entry.lastFetchTime) / 60_000.0
            log.log("ICS [${source.displayName}]: cache fresh (last fetch ${"%.1f".format(minAgo)} min ago) — skipping HTTP.")
            return@withContext entry.events
        }
        
        // Build conditional request
        val requestBuilder = Request.Builder().url(source.url)
        entry.eTag?.let { requestBuilder.addHeader("If-None-Match", it) }
        entry.lastModified?.let { requestBuilder.addHeader("If-Modified-Since", it) }
        
        log.log("ICS [${source.displayName}]: HTTP GET (ETag: ${entry.eTag ?: "none"})")
        
        try {
            val response = httpClient.newCall(requestBuilder.build()).execute()
            response.use { resp ->
                // 304 Not Modified
                if (resp.code == 304) {
                    entry.consecutiveFailures = 0
                    entry.lastFetchTime = now
                    log.log("ICS [${source.displayName}]: 304 Not Modified — using ${entry.events.size} cached event(s).")
                    return@withContext entry.events
                }
                
                // Non-success
                if (!resp.isSuccessful) {
                    return@withContext handleFetchFailure(
                        entry, source.displayName,
                        Exception("HTTP ${resp.code} ${resp.message}")
                    )
                }
                
                val icsContent = resp.body?.string() ?: ""
                
                // Content-hash dedup (skip re-parse if identical)
                val hash = computeHash(icsContent)
                if (hash == entry.contentHash) {
                    entry.consecutiveFailures = 0
                    entry.lastFetchTime = now
                    captureResponseHeaders(resp, entry)
                    log.log("ICS [${source.displayName}]: 200 but content unchanged — using ${entry.events.size} cached event(s).")
                    return@withContext entry.events
                }
                
                // Parse new content
                log.log("ICS [${source.displayName}]: new content (${icsContent.length} chars) — parsing.")
                val parsed = parseIcsContent(icsContent, source.id)
                log.log("ICS [${source.displayName}]: parsed ${parsed.size} event(s).")
                
                entry.contentHash = hash
                entry.events = parsed
                entry.lastFetchTime = now
                entry.consecutiveFailures = 0
                entry.backoffUntil = 0L
                captureResponseHeaders(resp, entry)
                
                return@withContext parsed
            }
        } catch (ex: Exception) {
            return@withContext handleFetchFailure(entry, source.displayName, ex)
        }
    }
    
    private fun readFromFile(source: CalendarSource): List<MeetingEvent> {
        val cacheKey = source.filePath!!
        val entry = cache.getOrPut(cacheKey) { SourceCache() }
        
        val file = File(source.filePath)
        val lastWrite = file.lastModified()
        
        if (entry.events.isNotEmpty() && lastWrite == entry.fileLastWrite) {
            log.log("ICS [${source.displayName}]: file unchanged — using ${entry.events.size} cached event(s).")
            return entry.events
        }
        
        return try {
            log.log("ICS [${source.displayName}]: reading file.")
            val icsContent = file.readText()
            val parsed = parseIcsContent(icsContent, source.id)
            log.log("ICS [${source.displayName}]: parsed ${parsed.size} event(s) from file.")
            
            entry.events = parsed
            entry.fileLastWrite = lastWrite
            entry.contentHash = computeHash(icsContent)
            
            parsed
        } catch (ex: Exception) {
            log.log("ERROR reading ICS file: ${ex.javaClass.simpleName}: ${ex.message}")
            entry.events // return stale cache rather than empty
        }
    }
    
    private fun handleFetchFailure(entry: SourceCache, displayName: String, ex: Exception): List<MeetingEvent> {
        entry.consecutiveFailures++
        val stepIdx = minOf(entry.consecutiveFailures - 1, BACKOFF_STEPS.size - 1)
        val backoff = BACKOFF_STEPS[stepIdx]
        entry.backoffUntil = System.currentTimeMillis() + backoff
        
        log.log("ERROR fetching ICS [$displayName] (failure #${entry.consecutiveFailures}): " +
                "${ex.javaClass.simpleName}: ${ex.message}. " +
                "Back-off for ${backoff / 60_000} min.")
        
        return entry.events // return stale cache rather than empty
    }
    
    private fun captureResponseHeaders(response: okhttp3.Response, entry: SourceCache) {
        response.header("ETag")?.let { entry.eTag = it }
        response.header("Last-Modified")?.let { entry.lastModified = it }
    }
    
    private fun computeHash(content: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val hashBytes = digest.digest(content.toByteArray())
        return hashBytes.joinToString("") { "%02x".format(it) }
    }
    
    /**
     * Sanitizes RRULE lines where both COUNT and UNTIL are present.
     * RFC 5545 §3.3.10 forbids supplying both; real-world feeds do it anyway.
     * We keep UNTIL and drop COUNT.
     */
    private fun sanitizeIcsContent(icsContent: String): String {
        // Unfold RFC 5545 line continuations
        var unfolded = icsContent.replace(Regex("\\r?\\n[ \\t]"), "")
        
        // Remove COUNT from RRULE lines that have both UNTIL and COUNT
        unfolded = unfolded.replace(
            Regex("(?im)^(RRULE[^:]*:[^\\r\\n]*UNTIL=[^\\r\\n]*)(?:;COUNT=\\d+|COUNT=\\d+;?)([^\\r\\n]*)\$"),
            "$1$2"
        )
        unfolded = unfolded.replace(
            Regex("(?im)^(RRULE[^:]*:(?:[^\\r\\n]*?))COUNT=\\d+;([^\\r\\n]*UNTIL=[^\\r\\n]*)\$"),
            "$1$2"
        )
        
        return unfolded
    }
    
    /**
     * Parses raw ICS text and returns a list of [MeetingEvent].
     */
    fun parseIcsContent(icsContent: String, calendarSourceId: String): List<MeetingEvent> {
        if (icsContent.isBlank()) return emptyList()
        
        return try {
            val sanitized = sanitizeIcsContent(icsContent)
            val builder = CalendarBuilder()
            val calendar: Calendar = builder.build(StringReader(sanitized))
            val events = mutableListOf<MeetingEvent>()
            
            for (component in calendar.getComponents<VEvent>(Component.VEVENT)) {
                val dtStart = component.getProperty<DtStart>(Property.DTSTART) ?: continue
                
                val uidProperty = component.getProperty<net.fortuna.ical4j.model.property.Uid>(Property.UID)
                val uid = uidProperty?.value?.takeIf { it.isNotBlank() } ?: UUID.randomUUID().toString()
                
                val summary = component.getProperty<net.fortuna.ical4j.model.property.Summary>(Property.SUMMARY)?.value ?: "(No Title)"
                val description = component.getProperty<net.fortuna.ical4j.model.property.Description>(Property.DESCRIPTION)?.value
                val location = component.getProperty<net.fortuna.ical4j.model.property.Location>(Property.LOCATION)?.value
                
                val startInstant = dateToInstant(dtStart.date)
                val dtEnd = component.getProperty<DtEnd>(Property.DTEND)
                val endInstant = if (dtEnd != null) {
                    dateToInstant(dtEnd.date)
                } else {
                    startInstant.plus(Duration.ofHours(1))
                }
                
                // Check for recurring events
                val rrule = component.getProperty<RRule>(Property.RRULE)
                if (rrule != null) {
                    // For recurring events, expand occurrences over the next 30 days
                    val rangeStart = Instant.now()
                    val rangeEnd = rangeStart.plus(Duration.ofDays(30))
                    
                    try {
                        val recur = rrule.recur
                        val seed = dtStart.date
                        val period = net.fortuna.ical4j.model.Period(
                            net.fortuna.ical4j.model.DateTime(java.util.Date.from(rangeStart)),
                            net.fortuna.ical4j.model.DateTime(java.util.Date.from(rangeEnd))
                        )
                        val occurrences = component.calculateRecurrenceSet(period)
                        
                        for (occurrence in occurrences) {
                            val occStart = Instant.ofEpochMilli(occurrence.start.time)
                            val occEnd = Instant.ofEpochMilli(occurrence.end.time)
                            val occUid = "${uid}_${occStart.epochSecond}"
                            
                            events.add(MeetingEvent(
                                uid = occUid,
                                title = summary,
                                start = occStart,
                                end = occEnd,
                                description = description,
                                location = location,
                                calendarSourceId = calendarSourceId
                            ))
                        }
                    } catch (e: Exception) {
                        // If recurrence expansion fails, add the base event
                        log.log("Warning: Failed to expand recurrence for $uid: ${e.message}")
                        events.add(MeetingEvent(
                            uid = uid,
                            title = summary,
                            start = startInstant,
                            end = endInstant,
                            description = description,
                            location = location,
                            calendarSourceId = calendarSourceId
                        ))
                    }
                } else {
                    events.add(MeetingEvent(
                        uid = uid,
                        title = summary,
                        start = startInstant,
                        end = endInstant,
                        description = description,
                        location = location,
                        calendarSourceId = calendarSourceId
                    ))
                }
            }
            
            events
        } catch (ex: Exception) {
            log.log("ERROR parsing ICS content: ${ex.javaClass.simpleName}: ${ex.message}")
            emptyList()
        }
    }
    
    private fun dateToInstant(date: java.util.Date): Instant {
        return when (date) {
            is net.fortuna.ical4j.model.DateTime -> Instant.ofEpochMilli(date.time)
            is net.fortuna.ical4j.model.Date -> Instant.ofEpochMilli(date.time)
            else -> date.toInstant()
        }
    }
}
